docs/reference-guides/interactivity-api/core-concepts/using-typescript.md

```diff
@@ -0,0 +1,782 @@
+			state.counter += 1;
+		},
+	},
+} );
+```
+
+The inferred types for this store are:
+
+```ts
+const myStore: {
+	state: {
+		counter: number;
+		readonly double: number;
+	};
+	actions: {
+		increment(): void;
+		// This behaves like a regular async function.
+		delayedIncrement(): Promise< void >;
+	};
+};
+```
+
+This also means that you can use your async actions in external functions, and TypeScript will correctly use the async function types.
+
+```ts
+const someAsyncFunction = async () => {
+	// This works fine and it's correctly typed.
+	await actions.delayedIncrement( 2000 );
+};
+```
+
+When you are not inferring types but manually writing the types for your entire store, you can use async function types for your async actions.
+
+```ts
+type Store = {
+	state: {
+		counter: number;
+		readonly double: number;
+	};
+	actions: {
+		increment(): void;
+		delayedIncrement(): Promise< void >; // You can use async functions here.
+	};
+};
+```
+
+However, there are a couple of things to consider when using asynchronous actions.
+
+1.  Just like with the derived state, if the asynchronous action needs to return a value and this value directly depends on some part of the global state, TypeScript will not be able to infer the type due to a circular reference.
+
+    ```ts
+    const { state, actions } = store( 'myCounterPlugin', {
+    	state: {
+    		counter: 0,
+    	},
+    	actions: {
+    		*delayedReturn() {
+    			yield new Promise( ( r ) => setTimeout( r, 1000 ) );
+    			return state.counter; // TypeScript can't infer this return type.
+    		},
+    	},
+    } );
+    ```
+
+    In this case, just as we did with the derived state, we must manually type the return value of the generator.
+
+    ```ts
+    const { state, actions } = store( 'myCounterPlugin', {
+    	state: {
+    		counter: 0,
+    	},
+    	actions: {
+    		*delayedReturn(): Generator< uknown, number, uknown > {
+    			yield new Promise( ( r ) => setTimeout( r, 1000 ) );
+    			return state.counter; // Now this is correctly inferred.
+    		},
+    	},
+    } );
+    ```
+
+    That's it! Remember that the return type of a Generator is the second generic argument: `Generator< unknown, ReturnType, unknown >`.
+
+2.  When yielding inside a generator, TypeScript does not resolve the types of the promises directly because it doesn't know what the controller will return after processing the promise.
+
+    ```ts
+    const getAsyncNumber = async () => {
+    	return 1;
+    };
+
+    store( 'myCounterPlugin', {
+    	actions: {
+    		*getSomeNumber() {
+    			const n = yield getAsyncNumber(); // TypeScript doesn't resolve this type.
+    		},
+    	},
+    } );
+    ```
+
+    To fix this, you can combine the use of `yield*` with the `typed()` helper, which converts the promise type into an iterator that can be correctly inferred by TypeScript when used in conjunction with `yield*`.
+
+    _**Note**: The `typed` function will be included in WordPress 6.7._
```

Comment:
It's fine to land it earlier as:

```suggestion
    _**Note**: The `typed` function introduced in WordPress 6.7._
```

This is a common practice to avoid unnecessary updates after major releases.